[BE] rewrite ProcessGroupNCCLTest to be MultiProcess

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67705

This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Commit:
[BE] rewrite ProcessGroupNCCLTest to be MultiProcess

This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -196,36 +197,52 @@ def tearDown(self):
     )
     def test_init_no_gpus(self):
-        store = c10d.FileStore(self.file.name, self.world_size)
+        store = c10d.FileStore(self.file_name, self.world_size)
```
Shouldn't this still be self.file.name since this test is not a MultiProcessTestCase?

pritamdamania87 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -196,36 +197,52 @@ def tearDown(self):
-    MAIN_PROCESS_RANK = 0
+class ProcessGroupNCCLTest(MultiProcessTestCase):
+    def _create_process_group_nccl(self, store, rank, world_size, opts):
+        pg = c10d.ProcessGroupNCCL(store, self.rank, self.world_size, opts)
+        dist.barrier(group=pg)
+        return pg
```
Why not just use `init_process_group`? That way you don't need to have an explicit call to `dist.barrier` since the barrier is done internally.

wanchaol replied:
discussed offline: we will keep the current testing structures (test_c10d_nccl and test_c10d_gloo testing the low-level/backend barebone apis, distributed_test.py testing the user facing APIs), ideally we want to only test the user facing APIs, created a issue to follow up: https://github.com/pytorch/pytorch/issues/68088

pritamdamania87 reviewed (commented):
Awesome work! Looks like test_reduce_scatter_ops is failing on CI.

Commit:
Update on "[BE] rewrite ProcessGroupNCCLTest to be MultiProcess"


This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[BE] rewrite ProcessGroupNCCLTest to be MultiProcess"


This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[BE] rewrite ProcessGroupNCCLTest to be MultiProcess"


This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[BE] rewrite ProcessGroupNCCLTest to be MultiProcess"


This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[BE] rewrite ProcessGroupNCCLTest to be MultiProcess"


This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -235,45 +252,61 @@ def test_empty_tensors(self):
             opts.rootTensor = rootTensor
-            work = pg.broadcast(xs, opts)
-            work.wait()
+            fut = pg.broadcast(xs, opts).get_future()
+            fut.wait()
```
n00b question, can you directly call .wait()?

wanchaol replied:
yeah both should work, i was trying to make it consistent with test_c10d_gloo so that we can merge it next, but it needs some additional work, I will just revert it back to directly calling wait()

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -235,45 +252,61 @@ def test_empty_tensors(self):
+            expected_tensor = torch.empty([i + 1, i + 1]).fill_(i + 1)
+            xs = [torch.empty([i + 1, i + 1]).fill_(-1).cuda(device=device_idx) for device_idx in rank_to_GPU[self.rank]]
+
+            # test with multiple input tensors
```
If understand this correctly, this means we have multiple GPUs in one rank, right? If so, can we make the comment a little bit clearer?

pritamdamania87 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -206,8 +206,8 @@ def test_init_no_gpus(self):
     def _create_process_group_nccl(self, store, rank, world_size, opts):
+        # create nccl pg with opts
         pg = c10d.ProcessGroupNCCL(store, self.rank, self.world_size, opts)
-        dist.barrier(group=pg)
```
Don't we need the barrier here since we are not doing init_process_group?

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -282,66 +315,50 @@ def allreduce(tensors, op):
-                    tensors[i],
-                )
+            # TODO(#38095): Replace assertEqualIgnoreType. See issue #38095
+            ndev = float(self.world_size)
```
Nit: we can move it before previous `assertEqualIgnoreType` so that this can be reused.

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -524,75 +552,71 @@ def reduce_scatter(outputs, input_lists, op):
+
+        for gpu in local_device_ids:
+            tensor_lists.append([t.cuda(device=gpu) for t in input_per_gpu])
 
         reduce_scatter(output, tensor_lists, c10d.ReduceOp.SUM)
 
-        for i in range(self.num_gpus):
+        for i in range(num_gpus):
             expected = torch.tensor(
                 [
-                    float(self.num_gpus * (self.num_gpus - 1) / 2)
-                    + (virtual_rank + i) * virtual_world_size
-                ]
-            )
+                    float((1 + self.world_size) * self.world_size / 2)
+                    + self.world_size * self.rank
+                ])
+
+
```
I am ok with this change. Maybe we can update the comment above to reflect this change? Also, if we can differentiate the sum among different ranks, that will be great. (Just a nit, :) ).

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -524,75 +552,71 @@ def reduce_scatter(outputs, input_lists, op):
+        for i in range(num_gpus):
+            prod_val = self.rank + 1
+            for _ in range(self.world_size - 1):
+                prod_val = prod_val * (prod_val + 1)
```
This does not seem to look correct. `prod_val` increases not by one but has been multiplied. So if you change the `batch_size` to 4. This test will fail.

wanchaol replied:
I didn't see a `batch_size` defined in the test, am i missing something here?

fduwjj replied:
Oh, I mean if you change it to 4 GPU, this logic will fail because you are changing the prod_val.

For example if num_gpus = 4 and rank = 0. The expected `prod_val` should be `4! = 24` but this logic will get:
`6 * 7 = 42`.

wanchaol replied:
yeah make sense, thanks for pointing out! just updated with the correct formula.

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -625,11 +652,11 @@ def allreduce(tensors):
                 # TODO(#38095): Replace assertEqualIgnoreType. See issue #38095
                 self.assertEqualIgnoreType(
-                    torch.tensor([float(i * (i + 1) / 2)]), tensors_list[i - 2][j]
+                    torch.tensor([(j + 1) * self.world_size]), tensors_list[i - 1][j]
```
N00b question, why do we not compare against tensors in works?

wanchaol replied:
hmmm sorry what do you mean "tensors in works"?

fduwjj replied:
I mean the list `works` define in L648

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -203,29 +204,45 @@ def test_init_no_gpus(self):
     def test_empty_tensors(self):
-        store = c10d.FileStore(self.file.name, self.world_size)
-        pg = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+        store = c10d.FileStore(self.file_name, self.world_size)
+        pg = self._create_process_group_nccl(store, self.rank, self.world_size, self.opts())
+        rank_to_GPU = init_multigpu_helper(self.world_size, "nccl")
+        local_device_idx = rank_to_GPU[self.rank][0]
```
This code seems to be very repetitive in each test. Maybe we can put them somewhere inside an object?

fduwjj commented:
Thanks for doing this and I learned a lot from this PR about the usage of each collective and I also add some comments or n00b questions. Correct me if I understand it wrong.

Commit:
Update on "[BE] rewrite ProcessGroupNCCLTest to be MultiProcess"


This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[BE] rewrite ProcessGroupNCCLTest to be MultiProcess"


This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[BE] rewrite ProcessGroupNCCLTest to be MultiProcess"


This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[BE] rewrite ProcessGroupNCCLTest to be MultiProcess"


This PR rewrites ProcessGroupNCCLTest to be MultiProcessTestCase. It was originally written in a single process multi-GPU fashion, we change it to multi-process instead to align with other c10d tests.

Differential Revision: [D32113626](https://our.internmc.facebook.com/intern/diff/D32113626/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32113626/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

fduwjj reviewed (approved):
The change looks good to me, thanks!
